File: src/main/java/com/thatsnomoon/autofm/VoiceEventListener.java
package com.thatsnomoon.autofm;

import net.dv8tion.jda.core.entities.Member;
import net.dv8tion.jda.core.events.guild.voice.GuildVoiceJoinEvent;
import net.dv8tion.jda.core.events.guild.voice.GuildVoiceLeaveEvent;
import net.dv8tion.jda.core.events.guild.voice.GuildVoiceMoveEvent;
import net.dv8tion.jda.core.hooks.ListenerAdapter;
import org.apache.logging.log4j.Logger;

import java.util.List;
import java.util.stream.Collectors;

/**
 * @author Ben
 */

public class VoiceEventListener extends ListenerAdapter {

    private final AutoFM parent;
    private final Logger LOG = Main.LOG;

    VoiceEventListener(AutoFM parent) {
        this.parent = parent;
    }

    @Override
    public void onGuildVoiceLeave(GuildVoiceLeaveEvent voiceLeaveEvent) {
        LOG.trace("Guild voice leave event");
        List<Member> members = voiceLeaveEvent.getChannelLeft().getMembers().stream().filter(m -> !m.getUser().isBot() || m.getUser().getIdLong() == m.getJDA().getSelfUser().getIdLong()).collect(Collectors.toList());
        if (!members.contains(voiceLeaveEvent.getGuild().getSelfMember())) return;
        if (members.size() == 1  && !parent.getGuildAudioPlayer(voiceLeaveEvent.getGuild()).player.isPaused()) {
            LOG.trace("No other users in current voice channel, pausing.");
            parent.getGuildAudioPlayer(voiceLeaveEvent.getGuild()).player.setPaused(true);
        }
    }

    @Override
    public void onGuildVoiceJoin(GuildVoiceJoinEvent voiceJoinEvent) {
        LOG.trace("Guild voice join event");
        if (voiceJoinEvent.getMember().getUser().getIdLong() == voiceJoinEvent.getJDA().getSelfUser().getIdLong()) {
            if (!parent.USES_STREAM) {
                LOG.trace("Joined voice channel, starting random track.");
                parent.nextTrack(voiceJoinEvent.getGuild());
            }
            else {
                LOG.trace("Joined voice channel, starting stream.");
                parent.startStream(voiceJoinEvent.getGuild());
            }
            if (voiceJoinEvent.getChannelJoined().getMembers().stream().filter(m -> !m.getUser().isBot() || m.getUser().getIdLong() == m.getJDA().getSelfUser().getIdLong()).collect(Collectors.toList()).size() < 2 && !parent.getGuildAudioPlayer(voiceJoinEvent.getGuild()).player.isPaused()) {
                LOG.trace("No other users in current voice channel, pausing.");
                parent.getGuildAudioPlayer(voiceJoinEvent.getGuild()).player.setPaused(true);
            }
            return;
        }
        LOG.trace("Paused: " + String.valueOf(parent.getGuildAudioPlayer(voiceJoinEvent.getGuild()).player.isPaused()));
        List<Member> members = voiceJoinEvent.getChannelJoined().getMembers().stream().filter(m -> !m.getUser().isBot() || m.getUser().getIdLong() == m.getJDA().getSelfUser().getIdLong()).collect(Collectors.toList());
        if (!members.contains(voiceJoinEvent.getGuild().getSelfMember())) return;
        if (members.size() >= 2 && parent.getGuildAudioPlayer(voiceJoinEvent.getGuild()).player.isPaused()) {
            LOG.trace("User just joined the current voice channel, unpausing.");
            parent.getGuildAudioPlayer(voiceJoinEvent.getGuild()).player.setPaused(false);
            if (parent.USES_STREAM) {
                parent.startStream(voiceJoinEvent.getGuild());
            }
        }

    }

    @Override
    public void onGuildVoiceMove(GuildVoiceMoveEvent voiceMoveEvent) {
        LOG.trace("Guild voice move event");
        List<Member> outOfMembers = voiceMoveEvent.getChannelLeft().getMembers().stream().filter(m -> !m.getUser().isBot() || m.getUser().getIdLong() == m.getJDA().getSelfUser().getIdLong()).collect(Collectors.toList());
        List<Member> intoMembers = voiceMoveEvent.getChannelJoined().getMembers().stream().filter(m -> !m.getUser().isBot() || m.getUser().getIdLong() == m.getJDA().getSelfUser().getIdLong()).collect(Collectors.toList());
        if (!outOfMembers.contains(voiceMoveEvent.getGuild().getSelfMember()) && !intoMembers.contains(voiceMoveEvent.getGuild().getSelfMember())) {
            LOG.trace("Not current voice channel, returning");
            return;
        }
        LOG.trace("Paused: " + String.valueOf(parent.getGuildAudioPlayer(voiceMoveEvent.getGuild()).player.isPaused()));
        if (outOfMembers.size() == 1 && !parent.getGuildAudioPlayer(voiceMoveEvent.getGuild()).player.isPaused()) {
            LOG.trace("No other users in current voice channel, pausing.");
            parent.getGuildAudioPlayer(voiceMoveEvent.getGuild()).player.setPaused(true);
            return;
        }

        if (!intoMembers.contains(voiceMoveEvent.getGuild().getSelfMember())) {
            LOG.trace("Not current voice channel, returning.");
            return;
        }
        if (intoMembers.size() >= 2 && parent.getGuildAudioPlayer(voiceMoveEvent.getGuild()).player.isPaused()) {
            LOG.trace("User joined current voice channel, unpausing.");
            parent.getGuildAudioPlayer(voiceMoveEvent.getGuild()).player.setPaused(false);
            if (parent.USES_STREAM) {
                parent.startStream(voiceMoveEvent.getGuild());
            }
        }
    }
}

File: src/main/kotlin/com/thatsnomoon/autofm/VoiceEventListener.kt
/*
 * Copyright (c) 2017 Benjamin Scherer
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
 */

package com.thatsnomoon.autofm

import net.dv8tion.jda.core.entities.Member
import net.dv8tion.jda.core.entities.VoiceChannel
import net.dv8tion.jda.core.events.guild.voice.GenericGuildVoiceEvent
import net.dv8tion.jda.core.events.guild.voice.GuildVoiceJoinEvent
import net.dv8tion.jda.core.events.guild.voice.GuildVoiceLeaveEvent
import net.dv8tion.jda.core.events.guild.voice.GuildVoiceMoveEvent
import net.dv8tion.jda.core.hooks.ListenerAdapter


/**
 * @author Ben
 */

class VoiceEventListener(private val parent: AutoFM): ListenerAdapter() {

    override fun onGenericGuildVoice(event: GenericGuildVoiceEvent) {
        val guild = event.guild
        /* the members in whatever voice channel is relevant */
        val members: List<Member>
        when (event) {
            is GuildVoiceJoinEvent -> {
                members = event.channelJoined.members
            }
            is GuildVoiceLeaveEvent -> {
                members = event.channelLeft.members
            }
            is GuildVoiceMoveEvent -> {
                members = when {
                    event.channelJoined.members.contains(guild.selfMember) -> {
                        event.channelJoined.members
                    }
                    event.channelLeft.members.contains(guild.selfMember) -> {
                        event.channelLeft.members
                    }
                    else -> return
                }
            }
            else -> return
        }
        /* if the bot isn't in the voice channel the event happened in it doesn't matter */
        if (!members.contains(guild.selfMember)) return

        /* if the bot just joined a channel and it's not playing or it's paused, start playing and unpause */
        if (event is GuildVoiceJoinEvent && (!parent.isPlaying(guild) || parent.isPaused(guild)) && event.member.isSelf) {
            parent.next(guild)
            parent.unpause(guild)
        }

        /* if there aren't any other members in the voice channel and the bot isn't paused already, pause */
        if (members.size < 2 && !parent.isPaused(guild)) {
            parent.pause(guild)
        }


        /* if there are other members in the voice channel and it was paused, unpause */
        else if (members.size > 1 && parent.isPaused(guild)) {
            parent.unpause(guild)
        }
    }
}